### PR TITLE
Left and right shift for IR

### PR DESCRIPTION
**Pull request by kddnewton**

### CODE REVIEW

File: yjit/src/backend/x86_64/mod.rs [hunk 1]
```diff
@@ -164,6 +164,35 @@ impl Assembler
...
+                        },
+                        // We have to load memory operands to avoid corrupting them
+                        (Opnd::Mem(_) | Opnd::Reg(_), _) => {
+                            (asm.load(opnds[0]), opnds[1])
+                        },
+                        _ => (opnds[0], opnds[1])
+                    };
+
+                    asm.push_insn(op, vec![opnd0, opnd1], target, text, pos_marker);
+                },
```
Comment (maximecb) [inline yjit/src/backend/x86_64/mod.rs]: Are these instructions modifying the input/output in-place like on x86?

Comment (kddnewton) [reply yjit/src/backend/x86_64/mod.rs]: Yeah that's why we're splitting here, same as Add or Sub, etc.

File: yjit/src/backend/x86_64/mod.rs [hunk 2]
```diff
@@ -164,6 +164,35 @@ impl Assembler
...
                 },
+                Op::Shl | Op::Shr => {
+                    let (opnd0, opnd1) = match (opnds[0], opnds[1]) {
+                        // Instruction output whose live range spans beyond this instruction
+                        (Opnd::InsnOut { .. }, _) => {
+                            let idx = match original_opnds[0] {
+                                Opnd::InsnOut { idx, .. } => {
+                                    idx
+                                },
+                                _ => panic!("nooooo")
```
Comment (XrXr) [inline yjit/src/backend/x86_64/mod.rs]:
```suggestion
                                _ => unreachable!()
```

Smaller binary and just as descriptive :p

File: yjit/src/backend/x86_64/mod.rs [hunk 3]
```diff
@@ -164,6 +164,35 @@ impl Assembler
 
                     asm.push_insn(op, vec![opnd0, opnd1], target, text, pos_marker);
                 },
+                Op::LShift | Op::RShift | Op::URShift => {
```
Comment (maximecb) [inline yjit/src/backend/x86_64/mod.rs:169]:
```suggestion
                // These instructions modify their input operand in-place, so we
                // may need to load the input value to preserve it
                Op::LShift | Op::RShift | Op::URShift => {
```

**Review by maximecb — APPROVED** [review]
Very nice 👌

**Pull request merged**